qiskit/pulse/parameter_manager.py

```diff
@@ -0,0 +1,422 @@
+        # accessing to private member
+        node._Schedule__children = [(t0, self.visit(sched)) for t0, sched in node.instructions]
+
+        # update timeslots
+        for chan in copy(node.channels):
+            if isinstance(chan.index, ParameterExpression):
+                chan_timeslots = node._timeslots.pop(chan)
+                new_channel = self.visit_Channel(chan)
+
+                # Merge with existing channel
+                if new_channel in node.timeslots:
+                    sched = Schedule()
+                    sched._timeslots = {new_channel: chan_timeslots}
+                    node._add_timeslots(0, sched)
+                # Or add back under the new name
+                else:
+                    node._timeslots[new_channel] = chan_timeslots
+
+        self._update_parameter_manager(node)
+        return node
+
+    def visit_AlignmentKind(self, node: AlignmentKind):
+        """Assign parameters to block's ``AlignmentKind`` specification."""
+        new_parameters = tuple(self.visit(param) for param in node._context_params)
+        node._context_params = new_parameters
+
+        return node
+
+    # Mid layer: Assign parameters to instructions
+
+    def visit_Call(self, node: instructions.Call):
+        """Assign parameters to ``Call`` instruction.
+
+        .. note:: ``Call`` instruction has a special parameter handling logic.
+            This instruction separately keeps program, i.e. parametrized schedule,
+            and bound parameters until execution. The parameter assignment operation doesn't
+            immediately override its operand data.
+        """
+        new_table = copy(node.arguments)
+
+        for parameter, value in new_table.items():
+            if isinstance(value, ParameterExpression):
+                new_table[parameter] = self._assign_parameter_expression(value)
+        node._arguments = new_table
+
+        return node
+
+    def visit_Instruction(self, node: instructions.Instruction):
+        """Assign parameters to general pulse instruction.
+
+        .. note:: All parametrized object should be stored in the operands.
+            Otherwise parameter cannot be detected.
+        """
+        node._operands = tuple(self.visit(op) for op in node.operands)
+
+        return node
+
+    # Lower layer: Assign parameters to operands
+
+    def visit_Channel(self, node: channels.Channel):
+        """Assign parameters to ``Channel`` object."""
+        if isinstance(node.index, ParameterExpression):
+            new_index = self._assign_parameter_expression(node.index)
+
+            # validate
+            if not isinstance(new_index, ParameterExpression):
+                if not isinstance(new_index, int) or new_index < 0:
+                    raise PulseError('Channel index must be a nonnegative integer')
+
+            # return new instance to prevent accidentally override timeslots without evaluation
```

Comment:
this also worries me. it seems that any change within `qiskit.pulse` could lead to unexpected, uncaught bugs in this file

Reply:
Yes, but I just ported the original implementation here (I added the comment since this caused a tox failure when I didn't make a copy). I think we can change the logic of timeslots to be more robust.

Reply:
Basically, channel stored in the `timeslots` key and instruction shares the object id. We update instruction channel, this implicitly updates the timeslot key.